### PR TITLE
Meowhash

### PR DESCRIPTION
**Pull request by olevino**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added new hash function: meowHash128

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Functions/FunctionsHashing.h [hunk 1]
```diff
@@ -1369,6 +1370,35 @@ class FunctionURLHash : public IFunction
...
+{
+    static constexpr auto name = "meowHash128";
+    using ReturnType = UInt128;
+
+    static UInt128 apply(const char *s, const size_t len)
+    {
+        union {
+            __m128i m128;
+            UInt128 u128;
+        };
```
Comment (yakov-olkhovskiy) [inline src/Functions/FunctionsHashing.h:1383]: I believe this is type punning which is UB.

File: src/Functions/FunctionsHashing.h [hunk 2]
```diff
@@ -1369,6 +1370,35 @@ class FunctionURLHash : public IFunction
...
+        m128 = MeowHash(MeowDefaultSeed, len, const_cast<char *>(s));
```
Comment (yakov-olkhovskiy) [inline src/Functions/FunctionsHashing.h:1384]: I don't think const_cast is a good approach - we shouldn't rely on current implementation of the library. Your previous variant was better, but you can't use strdup here - 's' can contain zeroes ( `SELECT meowHash128('aaa\0bbb')` ). Consider something like:
```
std::string str(s, len);
m128 = MeowHash(MeowDefaultSeed, str.size(), str.data());
```

Comment (alexey-milovidov) [reply src/Functions/FunctionsHashing.h:1385]: It's not worth extra copying and (possibly) allocations.
Let's patch the library or use const_cast (which is ok - many C libraries are written without good constness).

Comment (yakov-olkhovskiy) [reply src/Functions/FunctionsHashing.h:1385]: I believe GCC and clang both support __attribute__((__may_alias__)) - I already used it in another hash

**Pull request closed without merging**